### PR TITLE
[COR-106] [COR-107] [COR-108] Add Latency Monitoring

### PR DESCRIPTION
**Pull request by gunnarsundberg**
- Propagate traces through context instead of spans so that we may pass baggage in addition to spans
- Add rpc latency to ethereum subscription
- Add latency observations to baggage that will be used to derive metrics
- Add metrics exporter
- Add histogram metrics for latency metrics

### CODE REVIEW

**Review by rukolahasser — CHANGES_REQUESTED** [review]
1. There are only 4 spans, we need 5. It's missing an end to end span.
2. How do I check the histogram in jaeger?

**Comment by gunnarsundberg** [comment]
> * There are only 4 spans, we need 5. It's missing an end to end span.
> * How do I check the histogram in jaeger?

e2e is derived from other spans and exported as a metric to the collector instead of having its own span. jaeger does not support otel metrics or any other metrics capabilities so you cannot see the histograms. if you check out port 9999 of the collector and go to `/metrics` endpoint, you will be able to see the raw data. if you want to visualize it, it will need to be exported to something like our prometheus/grafana instance.

File: chain/ethereum/subscription.go [hunk 1]
```diff
@@ -258,32 +259,55 @@ func (s *Subscription) subscribeLogs(ctx context.Context) {
...
+			rcvTime := time.Now()
+			// Add rcvTime to baggage as connector latency observation
+			spanCtx := monitor.NewLatencyBaggage(context.TODO(), monitor.LatencyConnectorKey, rcvTime)
+
+			// Connector latency span begin
 			tr := monitor.CreateTracer(monitor.DefaultTracerName)
-			_, span := monitor.StartSpan(
-				context.TODO(),
+
+			spanCtx, span := monitor.StartSpan(
```
[the remark this answers is not in the document]
Comment (gunnarsundberg) [reply chain/ethereum/subscription.go:270]: this is connector latency span. I have an rpc latency span that starts from block time

**Review by rm289 — APPROVED** [review]
normally I don't and won't approve things, but these are typos (misspellings and one mistake on usage of context.).  I'm moving them forward because no one is on and we cannot afford to lose a day of development until reviewers come online.

**Comment by rm289** [comment]
> 1. There are only 4 spans, we need 5. It's missing an end to end span.
> 2. How do I check the histogram in jaeger?

The end to end span can be derived by adding the spans together.  @gunnarsundberg asked me if this was acceptable. I told him deriving times by adding two or more span times is acceptable.  The math will be the same since its simple addition.  
Let's please move this forward, we are running out of time to get this through staging and into prod.

File: monitor/metric.go [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+		rpcLatency:    {name: "latency.rpc.histogram", description: "Latency from block time to connector reception"},
+		connLatency:   {name: "latency.connector.histogram", description: "Latency from connector reception to kafka produce"},
+		coreLatency:   {name: "latency.core.histogram", description: "Latency from kafka to streamserver"},
+		systemLatency: {name: "latency.system.histogram", description: "Latency from connector to streamserver"},
+		e2eLatency:    {name: "latency.e2e.histogram", description: "Latency from block time to streamserver"},
+	}
+)
+
+// Export latency metrics from baggage in ctx
+func ExportLatencyMetrics(ctx context.Context, meter metric.Meter, connName string, env string) {
```
[the remark this answers is not in the document]
Comment (gunnarsundberg) [reply monitor/metric.go:58]: added!

File: kafkautils/consumer.go [hunk 3]
```diff
@@ -20,7 +20,7 @@ type Consumer struct {
 }
 
 const (
-	spanName          = "kafka -> end"
+	spanName          = "kafka -> consumer"
```
[the remark this answers is not in the document]
Comment (gunnarsundberg) [reply kafkautils/consumer.go:23]: hmu with a suggestion and I will change it after we push to staging

File: chain/ethereum/subscription.go [hunk 4]
```diff
@@ -62,10 +63,10 @@ type Subscription struct {
 
 type Log struct {
 	types.Log
-	Span trace.Span
+	Context context.Context
```
[the remark this answers is not in the document]
Comment (gunnarsundberg) [reply chain/ethereum/subscription.go:66]: ross wants me to push to staging so I will mark as resolved and address this after I push to staging. can we discuss on slack?

**Pull request merged**